### ISSUE TITLE
Avoid unnecessary update / re-render

### ISSUE THREAD

**Issue by tiago-sb** [livewire/livewire#712]
**Is your feature request related to a problem? Please describe.**

Everytime a component renders, all its DOM needs to be generated and, in cases where this generation is heavy, we could be spending time (cpu/user) unnecessarily.

Concrete example:
I have a search form that, depending on the inputs, can query the DB for 0.5 million records. (these are then paginated of course). I want this search to happen only when the user presses the "submit" button on the form.
With the current implementation, there is no way (that I know of!) to avoid a re-render when any of the wired inputs change (wire:model.lazy is still "too much").

Similar issues have been raised that maybe could maybe benefit from a common solution:
- [#575](https://github.com/livewire/livewire/issues/575)
- [#313](https://github.com/livewire/livewire/issues/313)
- [#494](https://github.com/livewire/livewire/issues/494)

**Describe the solution you'd like**

Have a modifier like 'wire:model.passive' (i'm not good at naming things :)) that prevents a syncInput action from being sent to the server, but instead remains in the actionQueue until another action (that does not have '.passive') is queued. At this time both actions are sent to the server (maybe in 1 single network request?).

**Describe alternatives you've considered**

1. Issue #313 solution would work here, although we'd have to manually manage the form data when the "livewire-way" would be to have the inputs wired and the component public variables automatically hydrated.

**Additional context**

Please note I've started using livewire 1 week ago, so please forgive me if this can be solved in a simple way!

**Comment by austenc** [comment]
Just ran into this as well. For this component I'm searching a bunch of text files, and when using `wire:model` the component re-renders too often for it to be performant. 

I ended up using the "hacky" solution Caleb mentioned in #313 for now.
```
<form wire:submit.prevent="submit(Object.fromEntries(new FormData($event.target)))">
```
Although this _will work_, a built-in solution would make me feel a lot better about the stability of this code... The question is, can this be implemented in a simplistic way that doesn't add much complexity to the framework? Could see either of these two approaches working:

### Option A
Add a modifier for `wire:model`. Similar to the proposed `.passive` option above, a bit different. What if (instead of queuing the updates) the ajax request to update the `wire:model` data could update the public property, but it wouldn't trigger a full render? So in this case you'd still have the ajax request, but it would just be "silent". This _could_ work, but in some cases may lead to race conditions... for example, what if the submit handler runs before the public property is updated?

### Option B
Add a way to "opt in" to form data being transferred with the `submit` action. Perhaps this could look something like `wire:submit.prevent.withData` -- or even the `$formData` magic variable suggested in #313. 

Would be willing to spike out a PR, but curious to hear opinions from others in the community first. Seems like emulating a "regular ol form" is a use case people will want. How to achieve this sort of thing didn't seem immediately obvious to me. 

What would be the preferred approach? My preference would be **Option B**. Is there a way of doing this that would still support nested fields like `some[thing]` too?

**Comment by calebporzio** [comment]
Great input here. I agree this should be possible.

So, pretend we could do this: `wire:submit=“save($formData)”` and `$formData` would be an array of all the form data.

Would that solve our problems? Or would there still be a need for something like `wire:model.passive`?

**Comment by austenc** [comment]
That solution would work fine for my purposes. Totally okay with manually updating the properties within the submit (`save` in your example) function. For reference, here's what the resultant function looked like from yesterday:

```php
public function submit($data)
{
    $this->search = $data['search'] ?? '';
    $this->on = $data['on_date'] ?? '';
    $this->render();
}
```
_Perhaps this is bad practice to call `render` like that? _

**Comment by tiago-sb** [comment]
Hi @calebporzio 

In my opinion the `wire:submit=“save($formData)”` doesn't really solve these problems in a clean and elegant way.
It is what I'm currently using since I could not find another option but it requires extra work to keep things ... synchronized.

Take this example:
```
class Search extends Component
{
    use WithPagination;

    public $searchTerm;

    public function render()
    {
        $searchResults = $this->doSearch();

        return view($this->getView(), ['searchResults' => $searchResults]);
    }

    public function search(array $formData)
    {
        $this->fill($formData);
        $this->page = 1;
        $this->validate(...);
    }

    private function doSearch(): array
    {
        return Model::where('title', $this->searchTerm)
            ->get()
            ->toArray();
    }
}
```
and the blade:
```
<form wire:submit.prevent="search(Object.fromEntries(new FormData($event.target)))">
    <input type="text" name="searchTerm" :value="$searchTerm"/>
</form>
```

I have to:
- manually fill in the component class public properties
- manually fill in the `value` of each input in the blade file

I would propose something in the lines of the `wire:model.passive` to be honest.
An alternative would be to somehow be able to skip the re-render - maybe
```
public function render()
{
    return null;
}
```
which would result in skipping the DOM update for obvious reasons ,i.e., keep the "old" DOM, but it then seems kind of useless: why send a syncInput action at all if nothing is going to change on the frontend - it's not like the server keeps state, livewire will send it all again in the next post anyway. We could send them later, together, when they'll actually be used.

So, to summarize, imho, I'd go for the "passive" solution.

**Comment by roni-estein** [comment]
I totally agree with @tiago-sb, I had this problem with my first livewire app months ago. I’ve been on a inertia.js environment since but I have two new projects I’m trying to use livewire for. Just can’t find the time with covid and being in isolation with 3 small kids. But hopefully in the next few days I’ll get better at daddy daycare and have more time. But .passive is so online with the livewire premise and mindset of less js.

**Comment by austenc** [comment]
After reading your latest message @tiago-sb (great code examples too!), I agree that having something like `.passive` sure would be slick. Especially when compared to all the manual work that is necessary in the component and the corresponding blade view. 

However, at some point there has to be a balance between framework support and "stuff that should belong in developer land". Not to say it shouldn't be supported, but if it causes significant complexity with the rest of Livewire's core, the gain may not be worth the pain, so to speak. 

@tiago-sb if there was a way to prevent `wire:model` ajax requests from triggering a re-render, but otherwise they worked the same, would that work for your needs? Guessing it would be easier to support than trying to queue the updates (and keep things in order when synced later).

Would be happy with either solution, but agree with both of you that the _ideal_ solution would be another form of `wire:model`. Maybe we could call it `wire:model.silent`? 🤷‍♂️

**Comment by tiago-sb** [comment]
Hi @austenc,

**yep, that would work** for the scenarios I have in mind/read about 👌
Thumbs up for `silent` instead of `passive` as well!

... the only thing that still bothers me a bit is that in these cases livewire will still be sending requests for nothing (right?).

Couldn't somehow the `addAction` mechanism be updated to bail out in the silent cases?
Something like: (note that my js skills are toddler level!)

```
addAction(action) {
        if (this.prefetchManager.actionHasPrefetch(action) && this.prefetchManager.actionPrefetchResponseHasBeenReceived(action)) {
            const message = this.prefetchManager.getPrefetchMessageByAction(action)

            this.handleResponse(message.response)

            this.prefetchManager.clearPrefetches()

            return
        }

        this.actionQueue.push(action)
       
        // ########## SOMETHING LIKE THIS

        action.passive && return

         // ##########

        // This debounce is here in-case two events fire at the "same" time:
        // For example: if you are listening for a click on element A,
        // and a "blur" on element B. If element B has focus, and then,
        // you click on element A, the blur event will fire before the "click"
        // event. This debounce captures them both in the actionsQueue and sends
        // them off at the same time.
        // Note: currently, it's set to 5ms, that might not be the right amount, we'll see.
        debounce(this.fireMessage, 5).apply(this)

        // Clear prefetches.
        this.prefetchManager.clearPrefetches()
    }
```

Anyway, simply having `silent` would be an awesome addition!!

**Comment by tiago-sb** [comment]
Hi @austenc,

First, thumbs up for `silent` instead of `passive` !

I thought a bit more about this and this solution will not help in all cases.

Take this example:
- I have a search form that returns X rows/entries
- Each entry is a livewire component
- Each entry has a "showDetails" button that should query the DB for more data on that specific entry and display it.

However, I don't want to get the details **everytime** the user presses the "showDetails" button, only the first time.
My current solution for this is storing the details in a public variable but:

- I have to either `toArray()` or `json_encode()` the details - cannot have a Collection or any data class
- the details are coming back and forth on each ajax request.

```php
class ListEntry extends Component
{
    public $contract;
    public $details;

    public function mount(Contract $contract)
    {
        $this->contract = $contract;
    }

    public function render()
    {
        return view('livewire.contracts.list-entry');
    }

    public function showDetails()
    {
        if ($this->details === null) {
            $this->details = goToDbAndGetDetails()->toArray();
        }
    }
```

... and having this silent solution here doesn't really help (or I'm not seeing how it does).

If, like you suggest, when using a silent modifier, only the re-render will be skipped, then I'd propose to allow the developer to decide whether to render or not, on the component class instead of doing so on the blade.
Instead of going for:

`<input type="text" wire:model.silent="model" />`

go for allowing render() to return null and in that case skip the re-render.
Because then we could manage it like this:

```php
class ListEntry extends Component
{
    public $details;

    private $shouldRender = true;

    public function render()
    {
        return $this->shouldRender ? view('livewire.contracts.list-entry') : null;
    }

    public function showDetails()
    {
        if ($this->details === null) {
            $this->details = goToDbAndGetDetails()->toArray();
        } else {
            $this->shouldRender = false;
        }
    }

    public function updated($name, $value)
    {
        if ($name === "some property I want to be silent") {
            $this->shouldRender = false;
        }
    }
}
```

Let me know your thoughts...we'll get there! 😀

**Comment by adevade** [comment]
**+1 for this!** 

I would like to use this functionality when I have a form with many fields that doesn't have to be reactive. 

You just want to call a function on `submit` that receives all data from the form, process that data and refresh the component.

It would be the most simple way to setup ajax forms, without having to define your own endpoints and controllers.

**Comment by Luddinus** [comment]
+1, I was looking those approach, happy to see there is already a "issue".

My case is very "simple", don't need any field to be reactive, only need to show the validation errors without refreshing the page.

I know I could use something like turbolinks, the thing is I have a little form in a modal, so that would re-render the page (and current modal will be hidden of course).

Anyway, much appreciate to Caleb Porzio with this ideas of Livewire and Alpine (and the others)!

**Comment by maciejfikus** [comment]
Also +1

As for now I encapsulated livewire ajax form to a laravel component in order to hide hackish way (mentioned above) to send forms without tons of server requests, but it is far from perfect. 

I do love idea of custom directive, but would be happy even with magic variable for a time being.

Anyway, thanks for Alpine.js and Livewire!

**Comment by mblarsen** [comment]
Not directly related, but I wrote this PR #975. The use cases doesn't quite fit here, but I'm linking the issues because there may be an overlap in a solution (to reduce updates across the board).

**Comment by caneco** [comment]
**Here's a "quick and dirty" solution for this…**

On a given `<input>`:
```HTML
<input wire:model="email"/>
``` 

I've added a "tiny" debounce:
```HTML
<input wire:model.debounce.2000s="email"/>
```

**Comment by calebporzio** [comment]
The next major version will have a new modifier called ".defer" that will defer the persisting of wire:model data until an action is performed.

Closing this.

**Comment by caneco** [comment]
Super! Thanks @calebporzio

**Comment by isometriq** [comment]
```
$this->skipRender()
```
> Skip running the ->render() method for the current request. (Usually for performance reasons)
https://laravel-livewire.com/docs/2.x/reference

**Comment by smartnathan** [comment]
Hi @calebporzio, the defer modifier doesn't seems to work in an input form with type "file". I've had this challenge when implementing multiple file upload using Uploadify jquery pluggin.
